sw/host/tests/crypto/aes_gcm_nist_kat/src/main.rs

```diff
@@ -0,0 +1,231 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use arrayvec::ArrayVec;
+use clap::Parser;
+use std::fs;
+use std::time::Duration;
+
+use serde::Deserialize;
+
+use cryptotest_commands::commands::CryptotestCommand;
+use cryptotest_commands::aes_gcm_commands::{
+    CryptotestAesGcmOperation, CryptotestAesGcmData, CryptotestAesGcmOutput,
+};
+
+use opentitanlib::app::TransportWrapper;
+use opentitanlib::execute_test;
+use opentitanlib::test_utils::init::InitializeTest;
+use opentitanlib::test_utils::rpc::{UartRecv, UartSend};
+use opentitanlib::uart::console::UartConsole;
+
+#[derive(Debug, Parser)]
+struct Opts {
+    #[command(flatten)]
+    init: InitializeTest,
+
+    // Console receive timeout.
+    #[arg(long, value_parser = humantime::parse_duration, default_value = "10s")]
+    timeout: Duration,
+
+    #[arg(long, num_args = 1..)]
+    aes_gcm_json: Vec<String>,
+}
+
+#[derive(Debug, Deserialize)]
+struct AesGcmTestCase {
+    vendor: String,
+    test_case_id: usize,
+    mode: String,
+    operation: String,
+    key_len: usize,
+    key: Vec<u8>,
+    aad: Vec<u8>,
+    iv: Vec<u8>,
+    tag: Vec<u8>,
+    ciphertext: Vec<u8>,
+    plaintext: Vec<u8>,
+    result: bool,
+}
+
+const AES_GCM_CMD_MAX_MSG_BYTES: usize = 64;
+const AES_GCM_CMD_MAX_KEY_BYTES: usize = 32;
+const AES_GCM_CMD_MAX_IV_BYTES: usize = 128;
+const AES_GCM_CMD_MAX_AAD_BYTES: usize = 90;
+const AES_GCM_CMD_MAX_TAG_BYTES: usize = 16;
+
+fn run_aes_gcm_testcase(
+    test_case: &AesGcmTestCase,
+    opts: &Opts,
+    transport: &TransportWrapper,
+    fail_counter: &mut u32,
+) -> Result<()> {
+    log::info!(
+        "vendor: {}, mode: {}, operation: {}, key_len (in bits): {}, test case: {}",
+        test_case.vendor,
+        test_case.mode,
+        test_case.operation,
+        test_case.key_len,
+        test_case.test_case_id
+    );
+    let uart = transport.uart("console")?;
+
+    CryptotestCommand::AesGcm.send(&*uart)?;
+
+    assert!(
```

Comment:
Some of my earlier test harnesses added these checks, completely out of paranoia that UART would be messed up if we panicked after sending a partial test case. This turns out to have no negative side effects (aside from the test failing, of course), so another option I have used more recently is to add the error message in an `expect` call when creating the `ArrayVec` for each input value, rather than including an explicit length check. The other nice thing about it is that `ArrayVec`'s length is a generic parameter that can be extrapolated by the compiler, so we wouldn't need to copy the max data lengths from the uJSON type definitions. Up to you whether to make this change.

Reply:
Could you refer to any existing test harness that reflects your suggestion? I guess that would be better.

Reply:
You can look at what I do in `sw/host/tests/crypto/hash_kat/src/main.rs` on line 95:
```
CryptotestHashMessage {
        message: ArrayVec::try_from(test_case.message.as_slice()).unwrap(),
        message_len: test_case.message.len(),
    }
    .send(&*uart)?;
```
We create the `ArrayVec` with the message and just `unwrap()` it, which will panic if the input is too long. The `ujson_rust` Bazel rule automatically generates the `CryptotestHashMessage` type, where the `message` field is specified to be an ArrayVec with generic constant length equal to the size specified in the uJSON C header (`hash_commands.h`).

Reply:
If you want to keep the error message you had before, you can replace the `unwrap()` with `expect("<message>")`.

Reply:
Ah I see. Yes I saw that file but was not sure you are referring to this as there's no `expect` wording there.

Sounds like a reasonable suggestion. Updating it soon. Thanks!

Reply:
Yeah, sorry about the confusion there; I had forgotten I only used `unwrap()`.